File: C++/1. Two Sum/solution.cpp
#include <vector>
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> storage;
        for (int i = 0; i < nums.size(); ++i)
        {
            if (i != 0 && storage.find(target - nums[i]) != storage.end())
            {
                return {storage[target - nums[i]], i};
            }
            storage[nums[i]] = i;
        }
        return {};
    }

    void runTests()
    {
        vector<int> testCase1{4, 5, 1, 44, 666, 214214, 55, 4, 561};
        int target1 = 5;
        vector<int> solution1{0, 2};
        cout << (twoSum(testCase1, target1) == solution1 ? "Test1: OK" : "Test1: Failed") << endl;

        vector<int> testCase2{2, 7, 11, 15};
        int target2 = 9;
        vector<int> solution2{0, 1};
        cout << (twoSum(testCase2, target2) == solution2 ? "Test2: OK" : "Test2: Failed") << endl;

        vector<int> testCase3{3, 3};
        int target3 = 6;
        vector<int> solution3{0, 1};
        cout << (twoSum(testCase3, target3) == solution3 ? "Test3: OK" : "Test3: Failed") << endl;
    }
};

int main(int argc, char const *argv[])
{
    Solution solution;
    solution.runTests();
    return 0;
}

File: C#/1. Two Sum/Solution.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        Dictionary<int, int> dictionary = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; ++i)
        {
            if (i != 0 && dictionary.ContainsKey(target - nums[i]))
            {
                return new int[2] { dictionary[target - nums[i]], i };
            }

            dictionary[nums[i]] = i;
        }

        return new int[2];
    }

    private void RunTests()
    {
        List<(int[], int, int[])> tests = new List<(int[], int, int[])>
        {
            (new int[]{4, 5, 1, 44, 666, 214214, 55, 4, 561}, 5, new int[]{0,2}),
            (new int[]{2, 7, 11, 15}, 9, new int[]{0, 1}),
            (new int[]{3, 3}, 6, new int[]{0, 1})
        };

        for (int i = 0; i < tests.Count; ++i)
            Console.WriteLine(TwoSum(tests[i].Item1, tests[i].Item2).SequenceEqual(tests[i].Item3) ? $"Test{i}: OK" : $"Test{i}: Failed");
    }

    public static void Main()
    {
        Solution solution = new Solution();
        solution.RunTests();
    }
}